subsys/bluetooth/audio/aics.c

```diff
@@ -0,0 +1,824 @@
+			return BT_GATT_ERR(BT_AICS_ERR_GAIN_MODE_NOT_ALLOWED);
+		}
+		if (inst->state.gain_mode != BT_AICS_MODE_AUTO) {
+			inst->state.gain_mode = BT_AICS_MODE_AUTO;
+			notify = true;
+		}
+		break;
+	default:
+		return BT_GATT_ERR(BT_AICS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (notify) {
+		inst->state.change_counter++;
+
+		BT_DBG("New state: gain %d, mute %u, gain_mode %u, counter %u",
+		       inst->state.gain, inst->state.mute,
+		       inst->state.gain_mode, inst->state.change_counter);
+
+		bt_gatt_notify_uuid(NULL, BT_UUID_AICS_STATE,
+				    inst->service_p->attrs, &inst->state,
+				    sizeof(inst->state));
+
+		if (inst->cb && inst->cb->state) {
+			inst->cb->state(NULL, (struct bt_aics *)inst, 0,
+					inst->state.gain, inst->state.mute,
+					inst->state.gain_mode);
+		} else {
+			BT_DBG("Callback not registered for instance %p", inst);
+		}
+	}
+
+	return len;
+}
+
+#if defined(CONFIG_BT_AICS)
+static void aics_description_cfg_changed(const struct bt_gatt_attr *attr,
+					 uint16_t value)
+{
+	BT_DBG("value 0x%04x", value);
+}
+#endif /* CONFIG_BT_AICS */
+
+static ssize_t write_description(struct bt_conn *conn,
+				 const struct bt_gatt_attr *attr,
+				 const void *buf, uint16_t len, uint16_t offset,
+				 uint8_t flags)
+{
+	struct bt_aics_server *inst = attr->user_data;
+
+	if (len >= sizeof(inst->description)) {
+		BT_DBG("Output desc was clipped from length %u to %zu",
+		       len, sizeof(inst->description) - 1);
+		/* We just clip the string value if it's too long */
+		len = (uint16_t)sizeof(inst->description) - 1;
+	}
+
+	if (memcmp(buf, inst->description, len)) {
+		memcpy(inst->description, buf, len);
+		inst->description[len] = '\0';
+
+		bt_gatt_notify_uuid(NULL, BT_UUID_AICS_DESCRIPTION,
+				    inst->service_p->attrs, &inst->description,
+				    strlen(inst->description));
+
+		if (inst->cb && inst->cb->description) {
+			inst->cb->description(NULL, (struct bt_aics *)inst, 0,
+					      inst->description);
+		} else {
+			BT_DBG("Callback not registered for instance %p", inst);
+		}
+	}
+
+	BT_DBG("%s", log_strdup(inst->description));
+
+	return len;
+}
+
+#if defined(CONFIG_BT_AICS)
+static ssize_t read_description(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr, void *buf,
+				uint16_t len, uint16_t offset)
+{
+	struct bt_aics_server *inst = attr->user_data;
+
+	BT_DBG("%s", log_strdup(inst->description));
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset,
+				 &inst->description, strlen(inst->description));
+}
+
+/************************ PUBLIC API ************************/
+void *bt_aics_svc_decl_get(struct bt_aics *aics)
+{
+	CHECKIF(!aics) {
+		BT_DBG("NULL instance");
+		return NULL;
+	}
+
+	return aics->srv.service_p->attrs;
+}
+
+static void prepare_aics_instances(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(aics_insts); i++) {
+		aics_insts[i].service_p = &aics_service_list[i];
+	}
+}
+
+int bt_aics_init(struct bt_aics *aics, struct bt_aics_init *init)
```

Comment:
It might be a good ideia to rename the struct bt_aics_init to be different than the function name, or perhaps rename the function name to something like bt_aics_register as it appears to be registering service instances at runtime.

Btw, we probably need to make clear that both forms of registration seems to be available.

Reply:
@Vudentz Thanks, that was actually something that was done to VOCS but not AICS for some reason. 

Regarding the single register (i.e. just `bt_aics_init/register` instead of `bt_aics_init` and `bt_aics_register`) makes a lot of sense to me. There's no reason why these can't be done in the same call, unless we want to unregister the callbacks, but I doubt that will make sense to do. 

I'll merge the two, and also add a complementary update to VOCS (and VCS/MICS)